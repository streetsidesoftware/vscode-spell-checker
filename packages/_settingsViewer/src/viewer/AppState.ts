/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
import { observable, computed, makeObservable, action } from 'mobx';
import { Settings, ConfigTarget, LocaleId, SettingByConfigTarget, WorkspaceFolder, TextDocument, ConfigSource } from '../api/settings/';
import { normalizeCode, lookupCode } from '../iso639-1';
import { compareBy, compareEach } from '../api/utils/Comparable';
import { uniqueFilter } from '../api/utils';
import { Messenger } from '../api';
import { ConfigTargets, configTargets } from '../api/settings/settingsHelper';

type TabTargets = ConfigTarget | 'file' | 'dictionaries' | 'about';

export interface Tab {
    label: string;
    target: TabTargets;
}

const tabs: Tab[] = [
    { label: 'User', target: 'user' },
    { label: 'Workspace', target: 'workspace' },
    { label: 'Folder', target: 'folder' },
    { label: 'File', target: 'file' },
    { label: 'Dictionaries', target: 'dictionaries' },
    { label: 'About', target: 'about' },
];

export interface LanguageInfo {
    code: string;
    name: string;
    dictionaries: string[];
    enabled: boolean;
}

export interface LanguageConfig {
    languages: LanguageInfo[];
    inherited?: ConfigSource;
}

export interface LanguageConfigs extends SettingByConfigTarget<LanguageConfig> {}

export interface State {
    activeTabName: string;
    settings: Settings;
    tabs: Tab[];
    activeTab: Tab;
    languageConfig: LanguageConfigs;
}

export interface FoundInConfig<T> {
    value: Exclude<T, undefined>;
    source: ConfigSource;
}

type InheritedFromSource<T> = {
    value: Exclude<T, undefined>;
    source: ConfigSource;
};

export class AppState implements State {
    @observable _activeTabName = '';
    @observable _settings: Settings = {
        dictionaries: [],
        knownLanguageIds: [],
        configs: {
            user: { inherited: {}, locales: [], languageIdsEnabled: [] },
            workspace: { inherited: { locales: 'user', languageIdsEnabled: 'user' }, locales: [], languageIdsEnabled: [] },
            folder: { inherited: { locales: 'user', languageIdsEnabled: 'user' }, locales: [], languageIdsEnabled: [] },
            file: undefined,
        },
    };
    @observable debugMode: boolean = false;

    constructor(private messageBus: Messenger) {
        makeObservable(this);
    }

    @computed get activeTabName() {
        return this._activeTabName;
    }

    @computed get settings() {
        return this._settings;
    }

    @computed get tabs() {
        const hidden = new Set<TabTargets>(configTargets.filter((target) => !this.settings.configs[target]));
        if (this.workspaceFolders.length <= 1) {
            hidden.add(ConfigTargets.folder);
        }
        if (this.workspaceFolders.length < 1) {
            hidden.add(ConfigTargets.workspace);
        }
        if (!this.settings.activeFileUri) {
            hidden.add('file');
        }
        return tabs.filter((tab) => !hidden.has(tab.target));
    }

    @computed get activeTab() {
        return this.tabs.find((t) => t.label === this.activeTabName) || this.tabs[0];
    }

    @computed get languageConfig(): LanguageConfigs {
        const calcConfig = (target: ConfigTarget): LanguageConfig => {
            const config = this.settings.configs[target];
            const locales = config.locales; // todo: calc inheritance
            const inherited = config.inherited.locales;

            const infos = new Map<string, LanguageInfo>();

            const addLocalesToInfos = (locales: string[], dictionaryName: string | undefined) => {
                locales
                    .map(normalizeCode)
                    .map(lookupCode)
                    .filter(notUndefined)
                    .forEach((lang) => {
                        const { code, lang: language, country } = lang;
                        const name = country ? `${language} - ${country}` : language;
                        const found = this.isLocalEnabledEx(target, code);
                        const enabled = (found && found.value) || false;
                        const info: LanguageInfo = infos.get(name) || {
                            code,
                            name,
                            enabled,
                            dictionaries: [],
                        };
                        if (dictionaryName) {
                            info.dictionaries.push(dictionaryName);
                        }
                        infos.set(name, info);
                    });
            };
            if (locales) {
                addLocalesToInfos(locales, undefined);
            }
            this.settings.dictionaries.forEach((dict) => addLocalesToInfos(dict.locales, dict.name));

            return {
                languages: [...infos.values()].sort(
                    compareEach(
                        compareBy((info) => !info.dictionaries.length),
                        compareBy('name')
                    )
                ),
                inherited,
            };
        };

        return {
            user: calcConfig('user'),
            workspace: calcConfig('workspace'),
            folder: calcConfig('folder'),
        };
    }

    @computed get activeTabIndex(): number {
        const index = this.tabs.findIndex((t) => t.label === this.activeTabName);
        return index > 0 ? index : 0;
    }

    @computed get workspaceFolderNames(): string[] {
        const workspace = this.settings.workspace;
        const folders = (workspace && workspace.workspaceFolders) || [];
        return folders.map((f) => f.name);
    }

    @computed get activeWorkspaceFolder(): string | undefined {
        const folder = this.findMatchingFolder(this.settings.activeFolderUri);
        return folder && folder.name;
    }

    @computed get workspaceFolders(): WorkspaceFolder[] {
        const workspace = this.settings.workspace;
        return (workspace && workspace.workspaceFolders) || [];
    }

    @computed get activeFileUri(): string | undefined {
        return this.settings.activeFileUri;
    }

    @computed get documents(): TextDocument[] {
        const workspace = this.settings.workspace;
        return (workspace && workspace.textDocuments) || [];
    }

    @computed get documentSelection(): { label: string; value: string }[] {
        return this.documents.map((doc) => ({ label: doc.fileName, value: doc.uri }));
    }

    private findMatchingFolder(uri: string | undefined): WorkspaceFolder | undefined {
        return this.workspaceFolders.filter((f) => f.uri === uri)[0];
    }

    private findMatchingFolderByName(name: string | undefined): WorkspaceFolder | undefined {
        return this.workspaceFolders.filter((f) => f.name === name)[0];
    }

    @action actionSetLocale(target: ConfigTarget, locale: LocaleId, enable: boolean) {
        const localesCurrent = this.settings.configs[target].locales;
        const locales = (enable ? [...localesCurrent, locale] : localesCurrent.filter((v) => v !== locale)).filter(uniqueFilter());
        this.settings.configs[target].locales = locales;
        const uri = this.settings.activeFolderUri;
        this.messageBus.postMessage({ command: 'EnableLocaleMessage', value: { target, locale, enable, uri } });
    }

    @action actionSetDebugMode(isEnabled: boolean) {
        this.debugMode = isEnabled;
    }

    private isLocalEnabledEx(field: ConfigTarget, code: LocaleId): InheritedFromSource<boolean> {
        const config = this.settings.configs[field];
        const source = config.inherited.locales || field;
        const locales = config.locales;
        return {
            value: locales.map(normalizeCode).includes(code),
            source,
        };
    }

    @action actionActivateTabIndex(index: number) {
        const tab = this.tabs[index];
        if (tab) {
            this.actionActivateTab(tab.label);
        }
    }

    @action actionActivateTab(tabName: string) {
        this._activeTabName = tabName;
        this.messageBus.postMessage({ command: 'SelectTabMessage', value: this.activeTabName });
    }

    @action actionSelectFolder(folderName: string) {
        const folder = this.findMatchingFolderByName(folderName);
        const folderUri = folder && folder.uri;
        this.settings.activeFolderUri = folderUri;
        if (folderUri !== undefined) {
            this.messageBus.postMessage({ command: 'SelectFolderMessage', value: folderUri });
        }
    }

    @action actionSelectDocument(documentUri: string) {
        this.messageBus.postMessage({ command: 'SelectFileMessage', value: documentUri });
    }

    @action actionEnableLanguageId(languageId: string, enable: boolean, target?: ConfigTarget) {
        const fileConfig = this.settings.configs.file;
        if (!target && fileConfig?.languageId === languageId) {
            fileConfig.languageEnabled = enable;
        }
        const uri = (!target && fileConfig?.uri) || this.settings.activeFolderUri;
        this.messageBus.postMessage({ command: 'EnableLanguageIdMessage', value: { languageId, enable, target, uri } });
    }

    @action updateSettings(settings: Settings): Settings {
        this._settings = settings;
        return this._settings;
    }
}

function notUndefined<T>(a: T): a is Exclude<T, undefined> {
    return a !== undefined;
}
