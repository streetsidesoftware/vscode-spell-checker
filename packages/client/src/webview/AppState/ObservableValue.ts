import deepEqual from 'fast-deep-equal';
import { createDisposable, type DisposableHybrid as Disposable } from 'utils-disposables';

export interface ObservableValue<T> {
    value: T;
    set(v: T): void;
    update(u: (v: T) => T): void;
    subscribe(s: (v: T) => unknown): Disposable;
}
class Observable<T> implements ObservableValue<T> {
    private _value: T;
    private _subscriptions = new Set<(v: T) => unknown>();
    private _busy = false;
    constructor(value: T) {
        this._value = value;
    }

    get value() {
        return this._value;
    }

    set value(v: T) {
        this.set(v);
    }

    set(value: T) {
        // Do not update if the value has not changed.
        if (this._value === value || deepEqual(this._value, value)) return;
        this._value = value;
        this.notify();
        return;
    }

    subscribe(s: (v: T) => unknown): Disposable {
        const subscriptions = this._subscriptions;
        subscriptions.add(s);
        return createDisposable(() => subscriptions.delete(s));
    }

    update(u: (v: T) => T) {
        return this.set(u(this._value));
    }

    private notify() {
        if (this._busy) return;
        try {
            this._busy = true;
            for (const s of this._subscriptions) {
                s(this._value);
            }
        } finally {
            this._busy = false;
        }
    }
}
export function createStoreValue<T>(v: T): ObservableValue<T> {
    return new Observable(v);
}
