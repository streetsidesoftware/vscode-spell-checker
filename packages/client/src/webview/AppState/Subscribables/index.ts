export * from './createFunctions';
export * from './operators/index';
export { pipe } from './pipe';
export { rx } from './rx';
export type * from './Subscribables';
export { createSubscribableView } from './SubscribableView';
