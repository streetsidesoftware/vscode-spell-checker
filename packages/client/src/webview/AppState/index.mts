export { awaitSubscribable } from '../../Subscribables/helpers/awaitSubscribable.js';
export { calcDocSettings, getWebviewGlobalStore } from './store.mjs';
