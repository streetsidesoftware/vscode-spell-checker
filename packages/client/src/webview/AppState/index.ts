export { awaitSubscribable } from '../../Subscribables/helpers/awaitSubscribable';
export { calcDocSettings, getWebviewGlobalStore } from './store';
