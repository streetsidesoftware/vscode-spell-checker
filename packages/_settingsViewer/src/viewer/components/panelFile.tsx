import InputLabel from '@material-ui/core/InputLabel';
import ListItem from '@material-ui/core/ListItem';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemSecondaryAction from '@material-ui/core/ListItemSecondaryAction';
import ListItemText from '@material-ui/core/ListItemText';
import Link from '@material-ui/core/Link';
import MenuItem from '@material-ui/core/MenuItem';
import Select from '@material-ui/core/Select';
import IconCode from '@material-ui/icons/Code';
import IconDescription from '@material-ui/icons/Description';
import { toJS } from 'mobx';
import { observer } from 'mobx-react';
import * as React from 'react';
import { FileConfig } from '../../api/settings';
import { AppState } from '../AppState';
import { LinkOpenFile } from './link';
import { CsCheckBox as Checkbox, CsFormControl as FormControl, CsList as List } from './primitives';
import { SectionConfigFileList } from './sectionConfigFileList';
import { SectionDictionaries } from './sectionDictionaries';

@observer
export class PanelFile extends React.Component<{ appState: AppState }> {
    render(): JSX.Element {
        const appState = this.props.appState;
        const settings = appState.settings;
        const config = settings.configs.file;
        const languageId = config?.languageId || 'unknown';
        const languageEnabled = config?.languageEnabled;
        const name = config?.uri.replace(/.*\//, '') || 'File';
        const dictionaries = config?.dictionaries || [];
        const configFiles = config?.configFiles || [];
        const select = (elem: HTMLSelectElement) => elem && this.props.appState.actionSelectDocument(elem.value);
        const onClick = () => {
            this.enableLanguageId(!languageEnabled);
        };
        return (
            <div>
                <h2>File</h2>
                <FormControl>
                    <InputLabel id="select-file-label">File</InputLabel>
                    <Select
                        labelId="select-file-label"
                        id="select-file"
                        value={appState.activeFileUri}
                        onChange={(evt) => select(evt.target as HTMLSelectElement)}
                    >
                        {appState.documentSelection.map((item) => (
                            <MenuItem value={item.value} key={item.value}>
                                {item.label}
                            </MenuItem>
                        ))}
                    </Select>
                </FormControl>
                <h2>Settings</h2>
                <List>
                    <ListItem>
                        <ListItemIcon>
                            <IconDescription />
                        </ListItemIcon>
                        <ListItemText primary={<i>{name}</i>} secondary={secondaryFileMessage(config)} />
                    </ListItem>
                    <ListItem role="checkbox" onClick={onClick}>
                        <ListItemIcon>
                            <IconCode />
                        </ListItemIcon>
                        <ListItemText
                            primary={
                                <label htmlFor="checkbox-language-enabled">
                                    File Type: <i>{languageId}</i>
                                </label>
                            }
                        />
                        <ListItemSecondaryAction>
                            <Checkbox id="checkbox-language-enabled" checked={languageEnabled} onClick={onClick} />
                        </ListItemSecondaryAction>
                    </ListItem>
                </List>
                <SectionConfigFileList configFiles={configFiles}></SectionConfigFileList>
                <SectionDictionaries dictionaries={dictionaries} sectionTitle="Active Dictionaries"></SectionDictionaries>
                {appState.debugMode ? (
                    <div>
                        <pre>{JSON.stringify(toJS(config), null, 2)}</pre>
                    </div>
                ) : (
                    ''
                )}
            </div>
        );
    }

    enableLanguageId(enable: boolean): void {
        const config = this.props.appState.settings.configs.file;
        const languageId = config && config.languageId;
        if (languageId) {
            this.props.appState.actionEnableLanguageId(languageId, enable);
        }
    }
}

function* br(...frags: React.ReactFragment[]) {
    if (!frags.length) return;
    let index = 0;
    for (const f of frags) {
        if (index++) {
            yield <br />;
        }
        yield f;
    }
}

function isDefined<T>(t: T | undefined): t is T {
    return t !== undefined;
}

function secondaryFileMessage(config: FileConfig | undefined): React.ReactFragment | undefined {
    if (!config) return undefined;

    const { fileIsInWorkspace, fileIsExcluded, fileIsIncluded, fileEnabled, gitignoreInfo, blockedReason } = config;
    const gitignored = !!gitignoreInfo?.matched;

    const excludedBy =
        config.excludedBy
            ?.map((e) => e.configUri && LinkOpenFile({ uri: e.configUri, text: e.name ? `${e.name} - "${e.glob}"` : `"${e.glob}"` } || '*'))
            .filter(isDefined) ?? [];

    const linkGitignore = formatGitignoreLink(gitignoreInfo);

    const messages = [
        [formatGitignoreMsg(gitignoreInfo), gitignored],
        [formatBlockedMsg(blockedReason), !!blockedReason],
        [linkGitignore, gitignored],
        ['File is excluded', fileIsInWorkspace && fileIsIncluded && fileIsExcluded],
        ['File is NOT in `files` to be checked.', fileIsInWorkspace && !fileIsIncluded],
        ['File is NOT in the workspace and excluded.', !fileIsInWorkspace && fileIsExcluded],
        ['File is NOT in the workspace.', !fileIsInWorkspace && !fileIsExcluded && fileEnabled],
        ['File is NOT spell checked because it is not in the workspace.', !fileIsInWorkspace && !fileIsExcluded && !fileEnabled],
    ] as const;

    const msg = messages
        .filter(([_, m]) => m)
        .map(([m]) => m)
        .filter(isDefined);
    return msg ? <div>{[...br(...msg, ...excludedBy)]}</div> : undefined;
}

function formatGitignoreLink(gitignore: FileConfig['gitignoreInfo']) {
    if (!gitignore) return undefined;

    const { gitignoreFileUri: uri, gitignoreName: name, line } = gitignore;

    return LinkOpenFile({ uri: uri, text: name, line });
}

function formatGitignoreMsg(gitignore: FileConfig['gitignoreInfo']) {
    if (!gitignore) return undefined;

    const { line, glob } = gitignore;

    return (
        <span>
            File is exclude by <b>.gitignore</b>. Line: <b>{line}</b>, Glob: <b>{glob}</b>
        </span>
    );
}

function formatBlockedMsg(blockedReason: FileConfig['blockedReason']) {
    if (!blockedReason) return undefined;

    const { message, documentationRefUri } = blockedReason;

    return (
        <span>
            {message} {documentationRefUri ? <Link href={documentationRefUri}>More Info...</Link> : undefined}
        </span>
    );
}
