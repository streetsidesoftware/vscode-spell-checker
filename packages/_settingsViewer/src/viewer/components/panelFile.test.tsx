import * as React from 'react';
import { create } from 'react-test-renderer';
import { sampleAppState } from '../../test/fixtures/AppState';
import { PanelFile } from './panelFile';

describe('File Panel Verification', () => {
    it('tests the snapshot', () => {
        const appState = getSampleAppState();
        const panelRenderer = create(<PanelFile appState={appState}></PanelFile>).toJSON()!;
        expect(panelRenderer).toMatchSnapshot();
    });

    it('tests the handler', () => {
        const appState = getSampleAppState();
        const panelRenderer = create(<PanelFile appState={appState}></PanelFile>);
        expect(appState.settings.configs.file!.languageEnabled).toBe(true);
        panelRenderer.toTree()!.instance.enableLanguageId(false);
        expect(appState.settings.configs.file!.languageEnabled).toBe(false);
        expect(appState._postedMessages).toEqual([
            {
                command: 'EnableLanguageIdMessage',
                value: {
                    enable: false,
                    languageId: 'typescript',
                    target: undefined,
                    uri: 'file:///Users/cspell/projects/vscode-cspell-dict-extensions/extensions/dutch/src/extension.ts',
                },
            },
        ]);
    });

    function getSampleAppState() {
        return sampleAppState();
    }
});
