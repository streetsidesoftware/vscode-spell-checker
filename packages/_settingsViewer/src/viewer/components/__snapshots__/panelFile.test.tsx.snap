// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`File Panel Verification tests the snapshot 1`] = `
<div>
  <h2>
    File
  </h2>
  <div
    className="MuiFormControl-root WithStyles(ForwardRef(FormControl))-root-1"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
      data-shrink={true}
      id="select-file-label"
    >
      File
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
      onClick={[Function]}
    >
      <div
        aria-haspopup="listbox"
        aria-labelledby="select-file-label select-file"
        className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
        id="select-file"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="button"
        tabIndex={0}
      >
        vscode-cspell-dict-extensions/extensions/dutch/src/extension.ts
      </div>
      <input
        aria-hidden={true}
        className="MuiSelect-nativeInput"
        onAnimationStart={[Function]}
        onChange={[Function]}
        required={false}
        tabIndex={-1}
        value="file:///Users/cspell/projects/vscode-cspell-dict-extensions/extensions/dutch/src/extension.ts"
      />
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSelect-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
  <h2>
    Settings
  </h2>
  <ul
    className="MuiList-root WithStyles(ForwardRef(List))-root-2 MuiList-padding"
  >
    <li
      className="MuiListItem-root MuiListItem-gutters"
      disabled={false}
    >
      <div
        className="MuiListItemIcon-root"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M14 2H6c-1.1 0-1.99.9-1.99 2L4 20c0 1.1.89 2 1.99 2H18c1.1 0 2-.9 2-2V8l-6-6zm2 16H8v-2h8v2zm0-4H8v-2h8v2zm-3-5V3.5L18.5 9H13z"
          />
        </svg>
      </div>
      <div
        className="MuiListItemText-root MuiListItemText-multiline"
      >
        <span
          className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
        >
          <i>
            extension.ts
          </i>
        </span>
        <p
          className="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
        >
          <div />
        </p>
      </div>
    </li>
    <li
      className="MuiListItem-container"
    >
      <div
        className="MuiListItem-root MuiListItem-gutters MuiListItem-secondaryAction"
        disabled={false}
        onClick={[Function]}
        role="checkbox"
      >
        <div
          className="MuiListItemIcon-root"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M9.4 16.6L4.8 12l4.6-4.6L8 6l-6 6 6 6 1.4-1.4zm5.2 0l4.6-4.6-4.6-4.6L16 6l6 6-6 6-1.4-1.4z"
            />
          </svg>
        </div>
        <div
          className="MuiListItemText-root"
        >
          <span
            className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            <label
              htmlFor="checkbox-language-enabled"
            >
              File Type: 
              <i>
                typescript
              </i>
            </label>
          </span>
        </div>
      </div>
      <div
        className="MuiListItemSecondaryAction-root"
      >
        <span
          aria-disabled={false}
          className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-8 MuiCheckbox-root WithStyles(ForwardRef(Checkbox))-root-3 MuiCheckbox-colorSecondary WithStyles(ForwardRef(Checkbox))-colorSecondary-5 PrivateSwitchBase-checked-9 Mui-checked WithStyles(ForwardRef(Checkbox))-checked-7 MuiIconButton-colorSecondary"
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={null}
        >
          <span
            className="MuiIconButton-label"
          >
            <input
              checked={true}
              className="PrivateSwitchBase-input-11"
              data-indeterminate={false}
              id="checkbox-language-enabled"
              onChange={[Function]}
              type="checkbox"
            />
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
              />
            </svg>
          </span>
        </span>
      </div>
    </li>
  </ul>
  <div>
    <h2>
      Config Files
    </h2>
    <ul
      className="MuiList-root makeStyles-root-12 MuiList-padding"
    >
      <li
        className="MuiListItem-root MuiListItem-gutters"
        disabled={false}
      >
        <div
          className="MuiListItemIcon-root"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19.14 12.94c.04-.3.06-.61.06-.94 0-.32-.02-.64-.07-.94l2.03-1.58c.18-.14.23-.41.12-.61l-1.92-3.32c-.12-.22-.37-.29-.59-.22l-2.39.96c-.5-.38-1.03-.7-1.62-.94l-.36-2.54c-.04-.24-.24-.41-.48-.41h-3.84c-.24 0-.43.17-.47.41l-.36 2.54c-.59.24-1.13.57-1.62.94l-2.39-.96c-.22-.08-.47 0-.59.22L2.74 8.87c-.12.21-.08.47.12.61l2.03 1.58c-.05.3-.09.63-.09.94s.02.64.07.94l-2.03 1.58c-.18.14-.23.41-.12.61l1.92 3.32c.12.22.37.29.59.22l2.39-.96c.5.38 1.03.7 1.62.94l.36 2.54c.05.24.24.41.48.41h3.84c.24 0 .44-.17.47-.41l.36-2.54c.59-.24 1.13-.56 1.62-.94l2.39.96c.22.08.47 0 .59-.22l1.92-3.32c.12-.22.07-.47-.12-.61l-2.01-1.58zM12 15.6c-1.98 0-3.6-1.62-3.6-3.6s1.62-3.6 3.6-3.6 3.6 1.62 3.6 3.6-1.62 3.6-3.6 3.6z"
            />
          </svg>
        </div>
        <div
          className="MuiListItemText-root"
        >
          <span
            className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            <a
              className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
              href="file:///Users/cspell/projects/vscode-cspell-dict-extensions/extensions/dutch/cspell.json"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
            >
              dutch/cspell.json
            </a>
          </span>
        </div>
      </li>
      <li
        className="MuiListItem-root MuiListItem-gutters"
        disabled={false}
      >
        <div
          className="MuiListItemIcon-root"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19.14 12.94c.04-.3.06-.61.06-.94 0-.32-.02-.64-.07-.94l2.03-1.58c.18-.14.23-.41.12-.61l-1.92-3.32c-.12-.22-.37-.29-.59-.22l-2.39.96c-.5-.38-1.03-.7-1.62-.94l-.36-2.54c-.04-.24-.24-.41-.48-.41h-3.84c-.24 0-.43.17-.47.41l-.36 2.54c-.59.24-1.13.57-1.62.94l-2.39-.96c-.22-.08-.47 0-.59.22L2.74 8.87c-.12.21-.08.47.12.61l2.03 1.58c-.05.3-.09.63-.09.94s.02.64.07.94l-2.03 1.58c-.18.14-.23.41-.12.61l1.92 3.32c.12.22.37.29.59.22l2.39-.96c.5.38 1.03.7 1.62.94l.36 2.54c.05.24.24.41.48.41h3.84c.24 0 .44-.17.47-.41l.36-2.54c.59-.24 1.13-.56 1.62-.94l2.39.96c.22.08.47 0 .59-.22l1.92-3.32c.12-.22.07-.47-.12-.61l-2.01-1.58zM12 15.6c-1.98 0-3.6-1.62-3.6-3.6s1.62-3.6 3.6-3.6 3.6 1.62 3.6 3.6-1.62 3.6-3.6 3.6z"
            />
          </svg>
        </div>
        <div
          className="MuiListItemText-root"
        >
          <span
            className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            <a
              className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
              href="file:///Users/cspell/projects/vscode-cspell-dict-extensions/cspell.json"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
            >
              vscode-cspell-dict-extensions/cspell.json
            </a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div>
    <h2>
      Active Dictionaries
    </h2>
    <ul
      className="MuiList-root makeStyles-root-12 MuiList-padding"
    >
      <li
        className="MuiListItem-container"
      >
        <div
          className="MuiListItem-root MuiListItem-gutters MuiListItem-secondaryAction"
          disabled={false}
        >
          <div
            className="MuiListItemIcon-root"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M9.4 16.6L4.8 12l4.6-4.6L8 6l-6 6 6 6 1.4-1.4zm5.2 0l4.6-4.6-4.6-4.6L16 6l6 6-6 6-1.4-1.4z"
              />
            </svg>
          </div>
          <div
            className="MuiListItemText-root MuiListItemText-multiline"
          >
            <span
              className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
            >
              node
            </span>
            <p
              className="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
            >
              List of NodeJS terms.
            </p>
          </div>
        </div>
        <div
          className="MuiListItemSecondaryAction-root"
        >
          
        </div>
      </li>
      <li
        className="MuiListItem-container"
      >
        <div
          className="MuiListItem-root MuiListItem-gutters MuiListItem-secondaryAction"
          disabled={false}
        >
          <div
            className="MuiListItemIcon-root"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M9.4 16.6L4.8 12l4.6-4.6L8 6l-6 6 6 6 1.4-1.4zm5.2 0l4.6-4.6-4.6-4.6L16 6l6 6-6 6-1.4-1.4z"
              />
            </svg>
          </div>
          <div
            className="MuiListItemText-root MuiListItemText-multiline"
          >
            <span
              className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
            >
              npm
            </span>
            <p
              className="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
            >
              List of Top 500 NPM packages.
            </p>
          </div>
        </div>
        <div
          className="MuiListItemSecondaryAction-root"
        >
          
        </div>
      </li>
      <li
        className="MuiListItem-container"
      >
        <div
          className="MuiListItem-root MuiListItem-gutters MuiListItem-secondaryAction"
          disabled={false}
        >
          <div
            className="MuiListItemIcon-root"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M9.4 16.6L4.8 12l4.6-4.6L8 6l-6 6 6 6 1.4-1.4zm5.2 0l4.6-4.6-4.6-4.6L16 6l6 6-6 6-1.4-1.4z"
              />
            </svg>
          </div>
          <div
            className="MuiListItemText-root MuiListItemText-multiline"
          >
            <span
              className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
            >
              typescript
            </span>
            <p
              className="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
            >
              JavaScript and Typescript terms.
            </p>
          </div>
        </div>
        <div
          className="MuiListItemSecondaryAction-root"
        >
          
        </div>
      </li>
    </ul>
  </div>
  
</div>
`;
